From b9a0b269e40b4b0c820d8a9b11e2c8a8078a16cc Mon Sep 17 00:00:00 2001
From: Sumaiya Javed <sumaiya.javed@catalyst.net.nz>
Date: Tue, 9 Jun 2026 14:10:29 +1200
Subject: [PATCH] core_grades penalty: Add support for other penalty plugins

---
 grade/classes/penalty_manager.php | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/grade/classes/penalty_manager.php b/grade/classes/penalty_manager.php
index c4c7fb90..f146103f 100644
--- a/grade/classes/penalty_manager.php
+++ b/grade/classes/penalty_manager.php
@@ -61,6 +61,23 @@ class penalty_manager {
         return array_filter(explode(',', get_config('core', 'gradepenalty_enabledmodules')));
     }
 
+    /**
+     * Get list of enabled penalty plugins.
+     *
+     * @return array List of enabled penalty plugins.
+     */
+    public static function get_enabled_penality_plugins(): array {
+        $plugins = [];
+        // Iterate through all the penalty plugins to calculate the total penalty.
+        foreach (core_plugin_manager::instance()->get_plugins_of_type('gradepenalty') as $pluginname => $plugin) {
+            if (gradepenalty::is_plugin_enabled($pluginname)) {
+                $plugins[] = $pluginname;
+
+            }
+        }
+        return $plugins;
+    }
+
     /**
      * Enable the grade penalty feature for a module.
      *
@@ -251,7 +268,18 @@ class penalty_manager {
 
         // Show penalty indicator if penalty is greater than 0.
         if ($grade->is_penalty_applied_to_final_grade()) {
-            $indicator = new \core_grades\output\penalty_indicator(2, $grade);
+            foreach (core_plugin_manager::instance()->get_plugins_of_type('gradepenalty') as $pluginname => $plugin) {
+                if (\core\plugininfo\gradepenalty::is_plugin_enabled($pluginname)) {
+                    $classname = "\\gradepenalty_{$pluginname}\\output\\penalty_indicator";
+                    if (class_exists($classname)) {
+                        $indicator = new $classname(2, $grade);
+                        break;
+                    }
+                }
+            }
+            if (!isset($indicator)) {
+                $indicator = new \core_grades\output\penalty_indicator(2, $grade);
+            }
             $renderer = $PAGE->get_renderer('core_grades');
             return $renderer->render_penalty_indicator($indicator);
         }
-- 
2.43.0

