commit 4301dbc6
Author: Sumaiya Javed <sumaiya.javed@catalyst.net.nz>
Date:   Thu Oct 23 10:04:26 2025 +1300

    mod_assign: Add suport for gradepenalty_reattemptmaxscore

diff --git a/mod/assign/classes/penalty/helper.php b/mod/assign/classes/penalty/helper.php
index 3419aacc..1650d42d 100644
--- a/mod/assign/classes/penalty/helper.php
+++ b/mod/assign/classes/penalty/helper.php
@@ -41,10 +41,19 @@ class helper {
 
         // Get the assignment instance.
         $assign = new assign($context, $cm, $cm->course);
-
-        // Check if due date is set.
-        if (!$assign->get_instance()->duedate) {
-            return false;
+        $penaltytypes = \core_grades\penalty_manager::get_enabled_penality_plugins();
+        if ($penaltytypes) {
+            if (in_array('duedate', $penaltytypes)) {
+                // Check if due date is set.
+                if (!$assign->get_instance()->duedate) {
+                    return false;
+                }
+            } else if (in_array('reattemptmaxscore', $penaltytypes)) {
+                // Check if the maxattempts is not set to 1.
+                if (!$assign->get_instance()->maxattempts > 1) {
+                    return false;
+                }
+            }
         }
 
         // Check if the grade type is set to GRADE_TYPE_VALUE (grade 1 to 100).
@@ -135,8 +144,23 @@ class helper {
             'itemnumber' => 0,
         ]);
 
+        $penaltytypes = \core_grades\penalty_manager::get_enabled_penality_plugins();
+        if (empty($penaltytypes)) {
+            debugging("Penalty plugins not found.");
+            return;
+        }
+        if (in_array('reattemptmaxscore', $penaltytypes)) {
+            $initalvalue = 1;
+            // As attemptnumber starts from 0.
+            $finalvalue = (int)$assigngrade->attemptnumber + 1;
+        } else {
+            $initalvalue = $submissiondate;
+            $finalvalue = $duedate;
+        }
+
+
         // Apply penalty.
-        $container = penalty_manager::apply_grade_penalty_to_user($userid, $gradeitem, $submissiondate, $duedate);
+        $container = penalty_manager::apply_grade_penalty_to_user($userid, $gradeitem, $initalvalue, $finalvalue);
         if ($container->get_grade_before_penalties() == 0) {
             // There is no deduction applied to grade 0.
             $deductedpercentage = 0;
diff --git a/mod/assign/mod_form.php b/mod/assign/mod_form.php
index f5b71ef4..3e7522ec 100644
--- a/mod/assign/mod_form.php
+++ b/mod/assign/mod_form.php
@@ -147,6 +147,7 @@ class mod_assign_mod_form extends moodleform_mod {
         $options += array_combine(range(1, 30), range(1, 30));
         $mform->addElement('select', 'maxattempts', get_string('maxattempts', 'mod_assign'), $options);
         $mform->addHelpButton('maxattempts', 'maxattempts', 'assign');
+        $mform->disabledIf('maxattempts', 'recalculatepenalty', 'eq', '');
 
         $choice = new core\output\choicelist();
 
@@ -262,8 +263,16 @@ class mod_assign_mod_form extends moodleform_mod {
             $mform->addHelpButton('gradepenalty', 'gradepenalty', 'mod_assign');
             $mform->setDefault('gradepenalty', 0);
 
-            // Hide if the due date is not enabled.
-            $mform->hideIf('gradepenalty', 'duedate[enabled]');
+            $penaltytypes = \core_grades\penalty_manager::get_enabled_penality_plugins();
+            if ($penaltytypes) {
+                if (in_array('duedate', $penaltytypes)) {
+                    // Hide if the due date is not enabled.
+                    $mform->hideIf('gradepenalty', 'duedate[enabled]');
+                } else if (in_array('reattemptmaxscore', $penaltytypes)) {
+                    // Hide if the maxattempts is not set to 1.
+                    $mform->hideIf('gradepenalty', 'maxattempts', 'eq', 1);
+                }
+            }
 
             // Hide if the grade type is not set to point.
             $mform->hideIf('gradepenalty', 'grade[modgrade_type]', 'neq', 'point');
