diff --git a/mod/assign/locallib.php b/mod/assign/locallib.php
index c8b07261..d7514c93 100644
--- a/mod/assign/locallib.php
+++ b/mod/assign/locallib.php
@@ -2090,7 +2090,10 @@ class assign {
             // Add penalty indicator, icon only.
             $penaltyindicator = '';
             if ($deductedmark > 0) {
-                $gradegrade = new \grade_grade();
+                $gradegrade =  new \grade_grade([
+                                    'userid' => $userid,
+                                    'itemid' => $this->get_grade_item()->get_grade($userid)->itemid
+                                ], true);
                 $gradegrade->deductedmark = $deductedmark;
                 $gradegrade->overridden = $userid > 0 ? $this->get_grade_item()->get_grade($userid)->overridden : 0;
                 $penaltyindicator = \core_grades\penalty_manager::show_penalty_indicator($gradegrade);
@@ -5503,9 +5506,12 @@ class assign {
                         '',
                         $cangrade
                     );
+
                     // Display the penalty indicator next to the penalized grade, if applicable.
                     $penaltyindicator = \core_grades\penalty_manager::show_penalty_indicator(
                         new \grade_grade([
+                            'userid' => $user->id,
+                            'itemid' => $gradebookgrade->items[0]->id,
                             'deductedmark' => $gradebookgrade->deductedmark ?? 0,
                             'overridden' => $gradebookgrade->overridden ?? 0,
                         ], false)
@@ -5738,6 +5744,8 @@ class assign {
                 // Display the penalty indicator next to the penalized grade, if applicable.
                 $penaltyindicator = \core_grades\penalty_manager::show_penalty_indicator(
                     new \grade_grade([
+                        'itemid' => $this->get_grade_item()->id,
+                        'userid' => $grade->userid,
                         'deductedmark' => $deductedmark,
                         'overridden' => $userid > 0 ? $this->get_grade_item()->get_grade($userid)->overridden : 0,
                     ], false)
@@ -8007,7 +8015,10 @@ class assign {
         // Penalty indicator.
         $userassigngrade = $gradinginfo->items[0]->grades[$userid];
         if (isset($userassigngrade->grade)) {
-            $gradegrade = new \grade_grade();
+            $gradegrade =  new \grade_grade([
+                                'userid' => $userid,
+                                'itemid' => $gradinginfo->items[0]->id
+                            ], true);
             $gradegrade->deductedmark = $userassigngrade->deductedmark;
             $gradegrade->overridden = $userassigngrade->overridden;
             $penaltyindicator = \core_grades\penalty_manager::show_penalty_indicator($gradegrade);
